tests/src/LinkParser.cpp

```diff
@@ -173,3 +176,80 @@ TEST(LinkParser, doesntParseInvalidLinks)
         ASSERT_FALSE(p.result().has_value()) << input.toStdString();
     }
 }
+
+TEST(LinkParser, full)
+{
+    struct TestCase {
+        QString input;
+        std::optional<ParsedLink> expectedResult;
+    };
+
+    std::vector<TestCase> tests{
+        {
+            .input = "https://forsen.tv",
+            .expectedResult =
+                ParsedLink{
+                    .protocol = QStringLiteral(u"https://"),
+                    .host = QStringLiteral(u"forsen.tv"),
+                    .rest = QStringLiteral(u""),
+                    .source = "https://forsen.tv",
+                },
+        },
+        {
+            .input = "forsen.tv",
+            .expectedResult =
+                ParsedLink{
+                    .protocol = QStringLiteral(u""),
+                    .host = QStringLiteral(u"forsen.tv"),
+                    .rest = QStringLiteral(u""),
+                    .source = "forsen.tv",
+                },
+        },
+        {
+            .input = "forsen.tv/",
+            .expectedResult =
+                ParsedLink{
+                    .protocol = QStringLiteral(u""),
+                    .host = QStringLiteral(u"forsen.tv"),
+                    .rest = QStringLiteral(u"/"),
+                    .source = "forsen.tv/",
+                },
+        },
+        {
+            .input = "forsen.tv/commands",
+            .expectedResult =
+                ParsedLink{
+                    .protocol = QStringLiteral(u""),
+                    .host = QStringLiteral(u"forsen.tv"),
+                    .rest = QStringLiteral(u"/commands"),
+                    .source = "forsen.tv/commands",
+                },
+        },
+    };
+
+    for (const auto &[input, expectedResult] : tests)
+    {
+        LinkParser p(input);
+        ASSERT_EQ(p.result().has_value(), expectedResult.has_value())
+            << input.toStdString();
+        if (p.result().has_value())
+        {
+            ASSERT_EQ(p.result()->protocol, expectedResult->protocol)
+                << input.toStdString() << ": "
+                << p.result()->protocol.toString().toStdString()
+                << " != " << expectedResult->protocol.toString().toStdString();
+            ASSERT_EQ(p.result()->host, expectedResult->host)
+                << input.toStdString() << ": "
+                << p.result()->host.toString().toStdString()
+                << " != " << expectedResult->host.toString().toStdString();
+            ASSERT_EQ(p.result()->rest, expectedResult->rest)
+                << input.toStdString() << ": "
+                << p.result()->rest.toString().toStdString()
+                << " != " << expectedResult->rest.toString().toStdString();
+            ASSERT_EQ(p.result()->source, expectedResult->source)
+                << input.toStdString() << ": "
+                << p.result()->source.toStdString()
+                << " != " << expectedResult->source.toStdString();
+        }
+    }
+}
```

Comment:
Why are these tests separate? All of them have the property that `source == protocol + host + rest`. If the input should be printed in the assertion, this should be added in `Check::check`.

Reply:
When reading through the tests, this was not immediately clear to me. Having an explicit input & expected output from a test makes it clearer. I understand this does not hit any paths that weren't hit before.

If you want I can move these tests into a separate PR